### ISSUE TITLE
fetchComponentDataBeforeRender doesn't get access to session

### ISSUE THREAD

**Issue by Epochi** [choonkending/react-webpack-node#171]
Hey,

If user is already logged in and coming to the site, I'm unable to get user session data when using this

app/Containers/Votes.jsx
```
class Vote extends Component {
....
  static need = [
    fetchTopics
  ]
...
}
```
to call 
server/config/routes.js

`app.get('/topic', topics.all);`

server/controllers/topics.js
```
exports.all = function(req, res) {
   console.log(req.user) //returns undefined
   console.log(req) //returns shows empty passport object
};

```

I think it's because of fetchComponentDataBeforeRender, it dispatches the action from itself and not from the component, so it doesn't pass the user. Because I can see that passport deserealizing the user before the promise is called  and these console logs

```
'app.get('/topic', fiunction(req,res){
console.log(req.protocol + '://' + req.get('host') + req.originalUrl); 
//returns http://localhost:8080/top
})
  app.get('*', function (req, res, next) {
   //  this fires before /topic as it should
   // req.user returns the user
   // console.log(req.protocol + '://' + req.get('host') + req.originalUrl); //returns https://www.myaddress.com/
    App.default(req, res);
  });
```

Is there any way to see the session object in this situation?

Ultimately my goal is to detect if the user is logged in and if he is, perform topics query and check which topics he has already voted on, to change the color of upvote arrow. Which I'm able to do if calling fetchTopics from componentsWillMount() inside of Vote.jsx

**Comment by choonkending** [comment]
@Epochi Nice use case :+1: for checking the user.

> Ultimately my goal is to detect if the user is logged in and if he is
Don't we already do this [here](https://github.com/choonkending/react-webpack-node/blob/master/app/server.jsx#L63). You can access the redux store to check if the user is authenticated.

You could probably modify [this](https://github.com/choonkending/react-webpack-node/blob/master/app/actions/topics.js#L128) function to conditionally check if the user is already logged in?

Let me know if that helps

**Comment by Epochi** [comment]
@choonkending  hey, thanks for the answer!

You are correct, both of those correctly get the user session, but when fetchTopics() gets called from through fetchComponentDataBeforeRender, the axios promise comes from the server itself.

So User sends GET from his ip with his session data,cookies, and then fetchComponentDataBeforeRender sends a new axios GET from localhost and creates new request object which doesn't have req.user or req.session. So I'm unable to identify if the request was made from the authorized user.

I'm kind of afraid of letting fetchTopics get the data with only state, because it's an exposed function and someone could change their state and I wouldn't be able to verify without being able to check the cookie

Currently I'm trying to find a way to 'pass on' the request object from the initial req to the internal server call, that should be achieved through a proxy i think, not sure how that has to work yet.

**Comment by choonkending** [comment]
> I'm kind of afraid of letting fetchTopics get the data with only state, because it's an exposed function and someone could change their state and I wouldn't be able to verify without being able to check the cookie

Definitely a valid concern! Thanks for pointing that out.

I might have to look into this further as I do not immediately know what the solution for this is.

**Comment by Epochi** [comment]
@choonkending I think I have found a solution! 

inside  [app/server.jsx](https://github.com/choonkending/react-webpack-node/blob/master/app/server.jsx#L58)

```
 */
export default function render(req, res) {
    const history = createMemoryHistory();
    const authenticated = req.isAuthenticated();
    const store = configureStore({
      user: {
        authenticated: authenticated,
        isWaiting: false,
        message: '',
        isLogin: true
      }
    }, history);
    
    if(authenticated){
    
    const requestInterceptor = axios.interceptors.request.use(function (config) {
          config.headers['cookie'] = req.headers.cookie;
          return config;
       }, function (error) {
          return Promise.reject(error);
    });
    const responseInterceptor = axios.interceptors.response.use(function (response) {
            axios.interceptors.request.eject(requestInterceptor);
            axios.interceptors.response.eject(responseInterceptor);
            return response;
           }, function (error) {
            axios.interceptors.request.eject(requestInterceptor);
            axios.interceptors.response.eject(responseInterceptor);
            return Promise.reject(error);
      });  
    }
      
    const routes = createRoutes(store);

  match({ routes, location: req.url }, (error, redirectLocation, renderProps) => {
    ...
  });
}
```

As I understand the render function gets called only once, because later on all routing is handled solely by react-router. So axios creates interceptor only for requests that are called by statics need and on response the interceptors destroy themselves. (not sure if it would work if there's more static functions, because they would delete themselves after first response, maybe it would be better to add it inside another .then after fetchComponentDataBeforeRender )

So basically I'm just sending users cookie to the server again 

As far as I can tell there is no way for users to access this file and/or function, so this should be safe?

**Comment by Epochi** [comment]
Although there might be a big hole, if you do not call  `axios.interceptors.request.eject(requestInterceptor);`
 at all,one users cookie appears on initial server renders for users that aren't logged in... 

This is because i put the interceptor on global axios object, so I believe if there were several users server rendering at virtually the same time, their cookies would mix up. Don't really have a way to test this though.

I will try to create an api that creates axios instance and use it inside actions/topics.js and app/server.jsx instead of default axios. I wonder if it will use new instance per request or I will just mirror this.

**Comment by Epochi** [comment]
I solved it! #184 I believe with this I'll be able to scale easily using only react-router for routing and support initial server side rendering for routes that respond with slightly different data depending on user authentication.
